### ISSUE TITLE
Enable the use of different stylesheets (including custom)

### ISSUE THREAD

**Issue by gdanov** [asciidoctor/atom-asciidoc-preview#39]
I would like to be able to use more than one template and if possible be able to specify my own

**Comment by magi42** [comment]
Could we have more detailed instructions for how to configure a stylesheet in Atom settings? The Bower configuration was a bit beyond me.

**Comment by mojavelinux** [comment]
@magi42 My explanation was a proposal, not something that currently works. Development is still needed here.

**Comment by KangarooCreativeTeam** [comment]
I'm just writing to +1 this.  

I always work in a dark-ish room, with dark themes.  I HATE that the preview is a blinding white! MY EYES!  ARRRRG!  >_<

RELEVANT: Last week, Atom 1.1 was released, & it's Markdown preview features using the same theme the rest of the editor is currently configured to use (with the option to use the default white Github theme).  [[Press release > "Themed Markdown Preview"](http://blog.atom.io/2015/10/29/atom-1-1-is-out.html)]

I think the best solution is to default to using the same theme as the editor (with the option to specify a custom stylesheet coming down the line in future when ready...).
I think it'll look the best / most consistent at _all_ times (whatever theme I've currently chosen to use in the editor, from all my favourites). 

Thank you :)

--HEAD KANGAROO
[KangarooCreativeTeam.com](KangarooCreativeTeam.com)

**Comment by nicorikken** [comment]
I agree with @KangarooCreativeTeam The Markdown preview neatly conforms to the theme. As the Atom themes already provide a variety of colors and fonts, in particular the ones used to style Markdown, I would opt to use a style-sheet that uses these theme-based CSS stylesheets in rendering.
The (github flavored) Markdown package has 2 built-in stylesheets: one adopting to the theme style, one mimicking the GitHub online style. The preference can be set in the package configuration. I can see a similar setup for the AsciiDoc preview, where there are 2 built-in themes: one showing the default Asciidoctor preview, one rendering according to the theme CSS.

I haven't looked at the rendering internals, but I would assume 2 themes can be statically linked throughout the package if necessary.
Than it would mainly require someone to create the style relating to theme-based styles.

**Comment by mojavelinux** [comment]
Correct, we need to develop a new theme for Asciidoctor that ties into the (LESS) theme variables. One approach is to start from scratch, a clean room implementation. Another approach is to use the [asciidoctor-stylesheet-factory](https://github.com/asciidoctor/asciidoctor-stylesheet-factory) to produce a LESS stylesheet that has references to the theme variables.

Creating a theme for Asciidoctor is not a small task (but it's not impossible either). Like with the grammar, there are a lot of components to deal with.

The benefit of a clean room implementation for the Atom preview is that the stylesheet only has to focus on the content part (the embeddable HTML). A lot of the default stylesheet deals with a) a CSS reset across browsers, which we don't need in this case and b) styles for the framing around the content.

**Comment by nicorikken** [comment]
Wouldn't changing fonts and colors be sufficient?

**Comment by mojavelinux** [comment]
Just changing the fonts and colors is something that is (mostly) accomplished using the stylesheet factory.

I still think that a clean implementation is the most ideal path. There's a lot of unnecessary CSS that the default stylesheet provides in this context, and it's a bit roundabout to use Sass to build LESS output. Taking it one step at a time, you'd probably be able to make a decent "native" stylesheet in a few days time.

**Comment by mojavelinux** [comment]
It also allows us to skate around the open issue that the stylesheet factory still can't produce a stylesheet for the embeddable HTML (see https://github.com/asciidoctor/asciidoctor-stylesheet-factory/issues/18).

**Comment by nicorikken** [comment]
:+1:

**Comment by nicorikken** [comment]
I cannot get started with development because of the `atom-space-pen-views` not being included. Downloading and linking the repo didn't work either. So at this very moment I cannot even start working on this issue. Until somebody fixes the development track, I will hold off spending further time on this issue. 😢

**Comment by mojavelinux** [comment]
@Mogztter or @anthonny, do you know what's going on with atom-space-pen-views? Have you been able to get past it in your local checkout?

**Comment by Mogztter** [comment]
No, everything is working fine on my machine. I'm using `apm` command line to hack on Atom.

**Comment by nicorikken** [comment]
I forgot to call `apm install`, now it is working just fine, and I've already made my first stylesheet modifications. I'd like to transfer the sass (scss) components stylesheets from the stylesheet-factory to less, to have a good basis for supporting all features. I found this helpful overview http://www.nicoespeon.com/en/2013/06/convert-sass-project-into-less/

**Comment by nicorikken** [comment]
@mojavelinux How would you consider the option to migrate the stylesheet factory to LESS (which enables side-by-side comparisons on the same theme styles), and then create a separate stylesheet which bases itself on the syntax styles?

**Comment by nicorikken** [comment]
@mojavelinux @ldez I did [some early prototyping](https://github.com/nicorikken/asciidoctor-stylesheet-factory/commit/bfe436e2bf6a7a79387109f829d88059feed303c) on a LESS version of the stylesheet factory. Both syntaxes are quite new to me, so plenty of difficulties to port it. Especially the `components/_asciidoc.scss` is a pain in the ass, because of the use of logic, iterators, and inheritance. I need a better understanding of both styling language to perform a succesful port. Porting the other files is relatively straightforward.
Another issue is the inclusion of `foundation/components/*` and `compass/css3/*` which are both SCSS based. It seems functions are used (like emCalc), and so I cannot simply include rendered CSS files.
Maybe (hopefully) I will be able to make a rendered import of the `components/_asciidoc.scss` and `settings/_defaults.scss` at least for now.
It seems the current inheritance model is something like:
```
└─┬ asciidoctor
  ├── components/awesome-icons
  ├─┬ components/asciidoc
  │ ├── foundation/components/inline-lists
  │ ├── foundation/components/tables
  │ ├── foundation/components/panels
  │ ├── foundation/components/type
  │ └── foundation/components/grid
  ├─┬ settings/asciidoctor
  │ └─┬ settings/defaults
  │   ├── compass/css3/box-shadow
  │   ├── foundation/components/global
  │   └── normalize
  └── components/print
```
What do you think? Hybrid between rendered and LESS, go for a clean-room implementation as suggested earlier, or go LESS all the way?

**Comment by ldez** [comment]
Sass is more flexible et more powerful than LESS.

We don't need to convert SASS to LESS, we can process the SASS.

**Comment by nicorikken** [comment]
@ldez How would you create a solution for parsing SASS rather then LESS? Using Grunt maybe? The main reason for going with LESS is to integrate with the syntax definitions from the editor. I read somewhere that Atom only supported LESS natively, not SASS.

**Comment by nicorikken** [comment]
@ldez can you give some hints for a solution to compile the SASS on the spot, incorporating the syntax styles?

**Comment by nicorikken** [comment]
@ldez going by http://stackoverflow.com/questions/4436643/is-there-a-sass-js-something-like-less-js maybe something like [sass.js](https://github.com/medialize/sass.js/)?

**Comment by ldez** [comment]
nope, I thinking about [Node-sass](https://github.com/sass/node-sass)

**Comment by mojavelinux** [comment]
Sorry for the delay. I'll chime in later today.

**Comment by nicorikken** [comment]
Just to report back: after the node suggestion, I did some initial stuff, but I'm lacking the Javascript skills to pull this off. Maybe I'll give it another go someday.

**Comment by mojavelinux** [comment]
We could also consider pulling in some of the stylesheets from these alternative theme sets:

* https://github.com/darshandsoni/asciidoctor-skins
* https://github.com/nerk/asciidoctor-bs-themes

**Comment by waiwada** [comment]
Experimental implementation for the issue.
https://github.com/waiwada/atom-asciidoc-preview/commit/57644ce9cf04f9fcc0c3398042c54320e7fa1fd6 

I think we can also start from CSS, but LESS/SASS.

1. Generate CSSs with asciidoctor-stylesheet-factory
2. Replace text in CSSs "body" to ".embedded-body"
3. Place CSSs to "styles/skins" directory
4. And select CSS in the setting window

For generating single html file contains styles, <body> element must have class name.
ex. <body></body> -> <body class="embedded-body"></body>

**Comment by mojavelinux** [comment]
We really, really need the stylesheet factory to produce a proper embeddable stylesheet. I don't really like the idea of using hacks to scope the CSS after the fact. See https://github.com/asciidoctor/asciidoctor-stylesheet-factory/issues/18. The more we keep avoiding this problem, the more pervasive the hack becomes.

**Comment by damgot** [comment]
Hello,
Readme file says : "If you want to customize the apparance of the preview, you can specify your own stylesheet in the settings."

I try to add '-a stylesheet="<path_to_my_CSS"' in package settings/Default attributes. but it doesn't work. 

How can I customize the preview ? Thx for help :)

**Comment by dquesadacr** [comment]
Hi! Are there any news on customizing the preview? 
As @damgot  said, I tried to change the stylesheet without success.
A dark theme would be awesome!
Cheers!

**Comment by refaelsh** [comment]
> 
> 
> Hello,
> Readme file says : "If you want to customize the apparance of the preview, you can specify your own stylesheet in the settings."
> 
> I try to add '-a stylesheet="<path_to_my_CSS"' in package settings/Default attributes. but it doesn't work.
> 
> How can I customize the preview ? Thx for help :)

Just out of curiosity, did you tried it without the quotes?
